concurrency/ArtListFragment.java

```diff
@@ -0,0 +1,588 @@
+            // start prefetch once we've had a chance to handle all the bindView calls for the
+            // current visible items.
+            mHandler.postDelayed(mStartPrefetchRunnable, PREFETCH_DELAY_MILLIS);
+        }
+    }
+
+    protected void startPrefetch() {
+        // Don't start a prefetch if we have detached from the activity. This can happen because this method is called
+        // from a handler with a delay.
+        if (getActivity() != null && shouldPrefetchArt() && mPrefetchArtJob == Job.INVALID) {
+            mPrefetchArtJob = addJob(new PrefetchArtJob());
+        }
+    }
+
+    private class PrefetchArtJob extends Job {
+
+        ConcurrentLinkedQueue<String> mImageQueue = new ConcurrentLinkedQueue<String>();
+
+        public PrefetchArtJob() {
+            int count = 0;
+            int index = 0;
+            int first = Math.max(mLastVisible, mLastPrefetchPosition);
+
+            if (!AmazonApplication.onMainThread()) {
+                Log.error(TAG, "PrefetchArtJob not on main thread");
+                throw new IllegalArgumentException();
+            }
+            Cursor cursor = getCursor();
+            if (cursor != null) {
+                count = cursor.getCount();
+                int remaining = count - first;
+                if (remaining <= 0) {
+                    mPrefetchComplete = true;
+                } else {
+                    remaining = Math.min(remaining, CacheManager.getMaxPrefetchCount());
+                    index = cursor.getColumnIndexOrThrow(getArtColumnName());
+                    cursor.moveToPosition(first);
+                    mLastPrefetchPosition = first;
+                    while (remaining-- > 0) {
+                        final String artId = cursor.getString(index);
+                        if (TextUtils.isEmpty(artId) == false) {
+                            // make sure we skip over empty strings.
+                            mImageQueue.add(artId);
+                        }
+                        if (!cursor.moveToNext()) {
+                            remaining = 0; // short circuit if end of cursor
+                        }
+                    }
+                }
+            }
+        }
+
+        @Override
+        public int run() throws Exception {
+            prefetchArt(mImageQueue);
+            return RESULT_SUCCESS;
+        }
+    }
+
+    protected abstract String getArtColumnName();
+
+    protected abstract ImageLoaderFactory.ItemType getImageType();
+
+    protected boolean shouldPrefetchArt() {
+        // Only prefetch if not using SICS.
+        return true;
+    }
+
+    private void prefetchArt(ConcurrentLinkedQueue<String> queue) {
+        if (mPrefetchComplete) {
+            return;
+        }
+
+        int artSize = getAlbumArtSize();
+
+        CacheManager cache = CacheManager.getInstance();
+        while (!queue.isEmpty() && !mListScrolling) {
+            final String id = queue.poll();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `isEmpty()` and `poll()` may not be thread-safe at lines: **464 and 465**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `poll()` will get `null`. This `null` may be used at line: **466**.

**Fix**  
 Consider calling `poll()`, checking instead of your current check if the returned object is `null`, and only then using that object.